Generate HTML documentation using MkDocs (WIP)

nodiscc opened the pull request:
Work on https://github.com/shaarli/Shaarli/issues/312

MkDocs is a static site generator geared towards building project documentation.
Documentation source files are written in Markdown, and configured with a single YAML file.

 * http://www.mkdocs.org/
 * http://www.mkdocs.org/user-guide/configuration/

~~The first commit (69d45c3) of this PR implements this. Following commits just pull the latest markdown documentation from the wiki, removes pandoc-generated files and adds mkdocs-generated files.~~ Rebased.

A preview of the documentation can be found at https://nodiscc.github.io/Shaarli/
it was deployed using `mkdocs gh-deploy --message "update html doc" --remote-name nodiscc`

The page presents a few styling problems on my browser (excessively wide margin). We should investigate fixing this, as well as deploying to https://readthedocs.org/.

Currently all pages follow a flat structure in the sidebar, they should follow the current wiki pages layout instead (using http://www.mkdocs.org/user-guide/writing-your-docs/#multilevel-documentation).

ArthurHoaro commented on Makefile:
```diff
@@ -184,40 +184,15 @@ doxygen: clean
-	done;
+	python3 -m venv venv/
+	bash -c 'source venv/bin/activate; \
+	pip install mkdocs; \
```
I don't think you should include dependency installation in the makefile. I almost don't know anything about python, but I see at least 3 problems with this:

  * it will run the install command everytime, which isn't necessary
  * I think I've seen `pip3` on systems with python 2 as default
  * pip may require root access (but that may be the issue you address with the 2 previous lines).

So basically, it's system dependent.

This target requires `mkdocs` library and that's it, as it was with `pandoc`.

virtualtam replied:
> I don't think you should include dependency installation in the makefile
> it will run the install command everytime, which isn't necessary

Safer: Tox setup with pinned dependencies (requirements)

> `pip3` on systems with python 2 as default

This only applies to system-wide installations with a package manager; yet it is highly discouraged to `pip install` stuff system-wide as it may break/conflict managed packages. Best practice: use virtualenvs :)

> pip may require root access

The first two commands indeed create a Python virtualenv, which is a local, user-managed Python installation

ArthurHoaro commented on mkdocs.yml:
```diff
@@ -0,0 +1,6 @@
+site_name: Shaarli Documentation
```
This file should be ignore in release archives (`.gitattributes`).

nodiscc replied:
done

ArthurHoaro commented on tpl/default/page.footer.html:
```diff
@@ -1,7 +1,7 @@
   <strong><a href="https://github.com/shaarli/Shaarli">Shaarli</a></strong>
   - The personal, minimalist, super-fast, database free, bookmarking service by the Shaarli community
-  - <a href="doc/Home.html" rel="nofollow">Help/documentation</a>
+  - <a href="doc_html/index.html" rel="nofollow">Help/documentation</a>
```
I'd prefer to have single subfolder regarding the doc in the root folder. If you don't want to mix both of them, maybe we can split them in another subfolder (`doc/md` and `doc/html`).

nodiscc replied:
mkdocs doesn't support having both the source and build directories in the same place. But I can use the Makefile target to move files to the appropriate dir. What is preferable?

 * `doc/index.html`
 * `doc/html/index.html`

nodiscc replied:
done

nodiscc commented:
Instructions for deployment on readthedocs can be found [here](https://read-the-docs.readthedocs.io/en/latest/getting_started.html#in-markdown).

nodiscc commented:
> I didn't take a look, how does readthedocs integration work?

This patch so far only addresses _local_ HTML doc generation. It removes the pandoc dependency.

 - [ ] Edit https://github.com/shaarli/Shaarli/wiki/Release-Shaarli and remove pandoc requirement

For online hosting on readthedocs.io, no changes to the repo are needed, configuration takes place at https://readthedocs.org/dashboard/shaarli/. I've created the project and added @virtualtam as maintainer, @ArthurHoaro if you'd like to create an account there we would add you as well.

readthedocs build for this PR can be seen at https://shaarli.readthedocs.io/en/rtfd/

> I don't think you should include dependency installation in the makefile
> I think I've seen pip3 on systems with python 2 as default
> pip may require root access (but that may be the issue you address with the 2 previous lines).

Yes those are the problems addressed by using [virtualenv](http://docs.python-guide.org/en/latest/dev/virtualenvs/): _virtualenv is a tool to create isolated Python environments. virtualenv creates a folder which contains all the necessary executables to use the packages that a Python project would need._

So the only dependency would be `python3-virtualenv`. pip packages are installed into the virtualenv (no root required, doesn't mess with system-wide python packages), by the virtualenv's `pip` executable (no possible python2/3 confusion).

- [ ] Add python3-virtualenv to requirements

> it will run the install command everytime, which isn't necessary

This is true and can be avoided by moving `rm -r venv` to the `clean` target. Should I do it? Only running `make clean` manually or the full build would require package reinstallation. If your venv has not been destroyed it will only reuse installed packages (or upgrade them).

Take it as a disposable build environment.

> Should we use readthedocs or Github pages?

readthedocs, but gh-pages would be equally easy with `mkdocs gh-deploy`. Though it will increase the repo's size (new orphan branch with html docs).

ArthurHoaro commented:
Thanks for the explanations! 

Yes, I think it would be better not to have to download `mkdocs` every time (even if it doesn't make a big difference since we don't generate the docs that often), otherwise the current behaviour is fine. Also remember to *git ignore* the generated folder.

> readthedocs, but gh-pages would be equally easy with mkdocs gh-deploy. Though it will increase the repo's size (new orphan branch with html docs).

readthedocs does the job automatically, so that's fine. :)
I've created [an account](https://readthedocs.org/profiles/ArthurHoaro/).

virtualtam commented:
Nice! I've added both of you to the https://readthedocs.org/projects/python-shaarli-client/ project as well :)

nodiscc commented:
Will have a look at this ASAP

nodiscc commented on Makefile:
```diff
@@ -184,40 +184,16 @@ doxygen: clean
 doc: clean
 	@rm -rf doc
 	@git clone https://github.com/shaarli/Shaarli.wiki.git doc
+	mv doc/Home.md doc/index.md
```
done

nodiscc commented on Makefile:
```diff
@@ -184,40 +184,16 @@ doxygen: clean
-	done;
-
-htmldoc: authors doc htmlsidebar htmlpages
+	python3 -m venv venv/
```
Is that necessary? It adds dependencies, and documentation generation is already automated (Makefile for local doc, readthedocs of online docs)

virtualtam replied:
> Is that necessary? It adds dependencies, and documentation generation is already automated

we'll need to add a Travis environment to ensure the documentation is properly generated before a PR is merged

> Makefile for local doc, readthedocs of online docs)

contributors might prefer editing and submitting documentation PRs from the Github web interface :)

nodiscc commented:
> update MD files with the current wiki state
> cleanup Github-flavoured Markdown
> cleanup pandoc resources: remove related Makefile targets
> add base mkdocs resources:
>  Mangled links at https://shaarli.readthedocs.io/en/rtfd/Development/

Done, see above commits.


TODO:

- [ ] page structure matching the current Sidebar, see Pages and Navigation
- [ ] freeze the current wiki
- [ ] fix remaining Markdown errors
- [ ] define a new workflow for documentation contribution and management :)
- [ ] Documentation: Add python3-virtualenv to requirements, remove pandoc requirement
- [ ] Fix sidebar ordering
- [ ] Deploy to readthedocs

nodiscc commented:
@virtualtam @ArthurHoaro I have fixed mkdocs/rtfd builds, and rebased everything in a single commit.

Maybe this could be merged in the current state? We can tackle remaining tasks in a separate issue.

 * Building local documentation (`make doc_html`) works as expected.
 * readthedocs.io builds work as expected (latest build: http://shaarli.readthedocs.io/en/rtfd/)
 * Running raedthedocs.io builds must be done manually on https://readthedocs.org/projects/shaarli/builds/.

 - [ ] Once merged, the repo url for https://readthedocs.org/projects/shaarli/ must be updated to point to https://github.com/shaarli/Shaarli/
 - [ ] https://github.com/shaarli/Shaarli/wiki can be closed
 - [ ] Update README.md links pointing to the documentation
 - [ ] Merge README.md in the main documentation (not unlike https://github.com/quodlibet/quodlibet)
 - [ ] Automate readthedocs.io builds
 - [ ] Move remaining task items to another issue
